borg/repository.py

```diff
@@ -504,21 +506,30 @@ class SegmentFull(Exception):
     _commit = header_no_crc_fmt.pack(9, TAG_COMMIT)
     COMMIT = crc_fmt.pack(crc32(_commit)) + _commit
 
+    MAX_PENDING_WRITE_FDS = resource.getrlimit(resource.RLIMIT_NOFILE)[0] // 4
+
     def __init__(self, path, limit, segments_per_dir, capacity=90):
         self.path = path
         self.fds = LRUCache(capacity,
-                            dispose=lambda fd: fd.close())
+                            dispose=self.close_fd)
         self.segment = 0
         self.limit = limit
         self.segments_per_dir = segments_per_dir
         self.offset = 0
         self._write_fd = None
+        self._write_fds = deque(maxlen=self.MAX_PENDING_WRITE_FDS)
+        logger.debug('LoggedIO: max_pending_write_fds=%d', self.MAX_PENDING_WRITE_FDS)
 
     def close(self):
-        self.close_segment()
+        self.close_segment(sync=True)
         self.fds.clear()
         self.fds = None  # Just to make sure we're disabled
 
+    def close_fd(self, fd):
+        if hasattr(os, 'posix_fadvise'):  # only on UNIX
+            os.posix_fadvise(fd.fileno(), 0, 0, os.POSIX_FADV_DONTNEED)
+        fd.close()
```

Comment:
hmm, as this is new in your changeset, do i see it right that our previous code spoiled the cache when reading a lot of data from a repo?